### PR TITLE
[RLlib] Bump Gymnasium to 0.28.1

### PR DESCRIPTION
**Pull request by Rohan138**
## Related issue number

#34396 #32999 #31587

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [X] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by shrekris-anyscale — APPROVED** [review]
The Serve changes look good to me.

File: doc/source/ray-core/examples/plot_pong_example.ipynb [hunk 1]
```diff
@@ -31,7 +31,7 @@
     "To run the application, first install some dependencies.\n",
     "\n",
     "```bash\n",
-    "pip install gymnasium[atari] gym==0.26.2\n",
+    "pip install gymnasium[atari]==0.28.1\n",
```
Comment (ArturNiederfahrenhorst) [inline doc/source/ray-core/examples/plot_pong_example.ipynb:34]: I'm curious, why did this change?

File: rllib/BUILD [hunk 2]
```diff
@@ -2841,7 +2841,7 @@ py_test(
 )
 
 py_test(
-     name = "tests/test_supported_spaces_ppo_no_preproceesor_gpu",
+     name = "tests/test_supported_spaces_ppo_no_preprocessor_gpu",
```
Comment (ArturNiederfahrenhorst) [inline rllib/BUILD:2844]: I litteraly can not see the difference here. Is this a git error?

File: rllib/models/catalog.py [hunk 3]
```diff
@@ -350,7 +350,7 @@ def get_action_dist(
                 "Unsupported args: {} {}".format(action_space, dist_type)
             )
 
-        return dist_cls, dist_cls.required_model_output_shape(action_space, config)
+        return dist_cls, int(dist_cls.required_model_output_shape(action_space, config))
```
Comment (ArturNiederfahrenhorst) [inline rllib/models/catalog.py:353]: Why is this change required?

File: rllib/utils/spaces/space_utils.py [hunk 4]
```diff
@@ -369,9 +369,12 @@ def map_(elem, s):
             elif s.dtype != elem.dtype:
                 elem = elem.astype(s.dtype)
 
-        elif isinstance(s, int):
+        elif isinstance(s, int) or isinstance(s, np.int_):
```
Comment (ArturNiederfahrenhorst) [inline rllib/utils/spaces/space_utils.py:372]: Can you add a comment here on why we need to check for np.int_ and operate on np.float_?

**Review by ArturNiederfahrenhorst — APPROVED** [review]
Thanks a lot! I saw users asking about this so let's make sure we answer them with the good news 🙂

File: doc/source/ray-core/examples/plot_pong_example.ipynb (hunk 1 above)
Comment (Rohan138) [reply doc/source/ray-core/examples/plot_pong_example.ipynb:34]: gymnasium 0.26.2 did not come with ALE prepackaged I believe; that's why we were doing this as well as the workaround where we we using ALE environments from gym through the gymnasium->gym wrapper e.g. `gymnasium.make("GymV26Environment-v0", env_id="ALE/Pong-v5")`.

File: rllib/BUILD (hunk 2 above)
Comment (Rohan138) [reply rllib/BUILD:2844]: preproceesor vs preprocessor ;)

File: rllib/models/catalog.py (hunk 3 above)
Comment (Rohan138) [reply rllib/models/catalog.py:353]: This is from your original PR, but basically `gymnasium` has the dtype of Discrete spaces as `np.int64` now isntead of a plain python int, which requires us to force cast it here

File: rllib/utils/spaces/space_utils.py (hunk 4 above)
Comment (Rohan138) [reply rllib/utils/spaces/space_utils.py:373]: Done :)

File: rllib/BUILD (hunk 2 above)
Comment (kouroshHakha) [reply rllib/BUILD:2844]: NICE Catch maaaan!

Comment (kouroshHakha) [reply rllib/BUILD:2844]: Artur and I have the same visual capacity :) It's clear that we are old duuuuudes :)

**Review by kouroshHakha — APPROVED** [review]

File: rllib/BUILD (hunk 2 above)
Comment (ArturNiederfahrenhorst) [reply rllib/BUILD:2844]: lol

**Comment by kouroshHakha** [comment]
@Rohan138 We need to also launch the release tests on this PR. It's a tricky one. Ping me again if the pass rate is similar to what we have today. 
attach the build-kite link to this PR.

**Review by sven1977 — APPROVED** [review]
Looks good to me, too. Thanks @Rohan138 for this upgrade and all the small fixes you made in this PR!

File: python/requirements/ml/requirements_rllib.txt [hunk 5]
```diff
@@ -5,7 +5,7 @@
 # Atari
 # TODO(sven): Still needed for Atari (need to be wrapped by gymnasium as it does NOT support Atari yet)
 gym==0.26.2
```
Comment (ArturNiederfahrenhorst) [inline python/requirements/ml/requirements_rllib.txt:7]: @Rohan138 to be removed

File: rllib/env/wrappers/pettingzoo_env.py
Comment (elliottower) [inline rllib/env/wrappers/pettingzoo_env.py:1]: Make sure to remove the `self.render_mode` argument from `(render)`, as done in my other PR (rendering mode is set at initialization of the environment, doesn't take the argument anymore) https://github.com/ray-project/ray/pull/34696/files#diff-0ce4e43172221ba4bd276378df0a3ee1cfb64312930aa2c74bcc3774d2a7975b

**Comment by sven1977** [comment]
Hey @elliottower , thanks for your answer! The main blocker for us right now is the fact that the current combination of package versions also requires MuJoCo to be installed (which we are trying to avoid):

* gymnasium[atari]==0.28.1
* imageio==2.31.1
* ale_py==0.8.1
* mujoco-py<2.2,>=2.1
* pettingzoo==1.23.1
* supersuit==3.8.0
* tinyscaler==1.2.6
* shimmy==0.2.1  # <- let me try to switch this to 1.1.0 rn ...

**Comment by elliottower** [comment]
> Hey @elliottower , thanks for your answer! The main blocker for us right now is the fact that the current combination of package versions also requires MuJoCo to be installed (which we are trying to avoid):
> 
> * gymnasium[atari]==0.28.1
> * imageio==2.31.1
> * ale_py==0.8.1
> * mujoco-py<2.2,>=2.1
> * pettingzoo==1.23.1
> * supersuit==3.8.0
> * tinyscaler==1.2.6
> * shimmy==0.2.1  # <- let me try to switch this to 1.1.0 rn ...

I believe certain versions of mujoco lead to issues with certain gymnasium environments because of breaking changes on their end, pinged the main PMs for gymnasium to give a better explanation but there’s some specific versions that are required in pyproject (https://github.com/Farama-Foundation/Gymnasium/blob/a40234f5ef59aadfa6d6bad4cd0e05264868cb56/pyproject.toml#L44) but those shouldn’t be necessary for Atari so I’m not sure exactly why you’d be having issues.

File: docker/ray-ml/Dockerfile [hunk 6]
```diff
@@ -46,6 +51,14 @@ RUN sudo apt-get update \
         requirements*.txt \
     && sudo apt-get clean
 
+# MuJoCo Installation.
+ENV MUJOCO_GL=osmesa
```
Comment (araffin) [inline docker/ray-ml/Dockerfile:55]: if you use `v4` envs, you should be able to download the new bindings via the `mujoco` python package, it is much easier to install than `mujoco-py` + manual mujoco download.

Comment (araffin) [reply docker/ray-ml/Dockerfile:55]: to make it clear: `mujoco` python package doesn't require a manual install of mujoco anymore

File: .buildkite/pipeline.ml.yml [hunk 7]
```diff
@@ -530,15 +530,7 @@
   commands:
     - cleanup() { if [ "${BUILDKITE_PULL_REQUEST}" = "false" ]; then ./ci/build/upload_build_info.sh; fi }; trap cleanup EXIT
 
-    # Install mujoco necessary for the testing environments
```
Comment (sven1977) [inline .buildkite/pipeline.ml.yml:533]: Moved this logic to `install-dependencies.sh`

File: docker/ray-ml/Dockerfile (hunk 6 above)
Comment (sven1977) [reply docker/ray-ml/Dockerfile:55]: I did try that (replacing mujoco-py wih mujoco), but it still gave me a "MoJoCo not installed" error. So I reverted back to what I knew and used mujoco-py plus the tar download and extract procedure.

Comment (sven1977) [reply docker/ray-ml/Dockerfile:55]: Thanks for the hint @araffin ! We can do this in a follow up PR now that all tests are green (just want to keep it this way :) ).

File: doc/source/rllib/rllib-models.rst [hunk 8]
```diff
@@ -162,7 +162,7 @@ For example, for manipulating your env's observations or rewards, do:
 
 .. code-block:: python
 
-    import gym
+    import gymnasium as gym
```
Comment (sven1977) [inline doc/source/rllib/rllib-models.rst:165]: bug fix

**Review by maxpumperla — APPROVED** [review]

**Review by ArturNiederfahrenhorst — APPROVED** [review]

**Review by ericl — APPROVED** [review]
Stamping

**Comment by ArturNiederfahrenhorst** [comment]
Another related issue: https://github.com/ray-project/ray/issues/37393

**Comment by ArturNiederfahrenhorst** [comment]
@sven1977 I'm adding a PR ontop of this to revamp our video-recording capabilities: https://github.com/ray-project/ray/pull/37922

**Comment by XavierGeerinck** [comment]
Thank you so much for this PR! Can't wait for it to get merged as we are heavily waiting on this as well. Just a quick question, would it be feasible to go straight to the latest version then as well? (0.29.0 is out meanwhile)

**Comment by ChrisMcCarthyDev** [comment]
Thank you for the effort that's went into this! Is there a proposed release date?

**Comment by ArturNiederfahrenhorst** [comment]
@ChrisMcCarthyDev Around three weeks.
